### ISSUE TITLE
Notebooks are loaded from another instance if multiple instances are hosted on the same domain

### ISSUE THREAD

**Issue by leukipp** [jupyterlite/jupyterlite#440]
## Description

Hosting multiple JupyterLite instances on the same domain (e.g. GitHub pages) leads to a weird behavior for equally named files.
Completely different .ipynb files from instance A will be loaded when you afterwards visit instance B.

## Reproduce

1. Go to project [https://leukipp.github.io/jupyterlite/lab/](https://leukipp.github.io/jupyterlite/lab/), forked from the demo repository.
2. Open `python.ipynb`, make some changes and save.
3. Go to another project [https://leukipp.github.io/xmastree/lab/](https://leukipp.github.io/xmastree/lab/), which also contains a file named `python.ipynb`.
4.  Open `python.ipynb`, the content from the previous visited notebook is loaded.

## Expected behavior

Notebook states should be loaded correctly, regardless of the previously visited instances.

## Context

- JupyterLite version: v0.1.0a18

## Possible Solutions

- Use different indexedDB database names per JupyterLite root path (currently always `JupyterLite Storage` is used) 
- Use relative filepaths inside indexedDB database (e.g. /jupyterlite/lab/... or /xmastree/lab/...)

**Comment by bollwyvl** [comment]
I feel like in many cases (and probably the default), letting user-created data persist between different apps, deployments on the same domain is probably fine... we generally try to err on the side of giving the _end user_ more power unless it specifically conflicts with the ability to create a stable system. Certainly, if two apps are deployed that have different extensions that would make some notebooks (for example) not work at all would require doing something directly about this.

> different indexedDB database names per JupyterLite

A few options:
- make `appName` drive the [`STORAGE_NAME`](https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/contents.ts#L16)
- have a new key, e.g. `localStorageName`
- some combination of the two (prefer `localStorageName`, fall back to `appName`)

That last option sounds good, as it would keep the existing behavior, by default, allow changing the `appName` to get some isolated content (and tabs, etc), and allowing multiple apps to share the same contents.

> relative filepaths inside indexedDB

Relative paths are pain: a lot of our complexity in e.g. `config-utils.js` that i'd _love_ to get rid of is around trying to _avoid_ the apps knowing anything about relative paths... it's just very hard to predict how they will be resolved in the face of trailing slashes, etc. on different hosts.

**Comment by leukipp** [comment]
> I feel like in many cases (and probably the default), letting user-created data persist between different apps, deployments on the same domain is probably fine...

...for data on a settings level this sounds good. For data on a filesystem level it could cause confusions, at least for the scenario I mentioned above.

Another option:
> A few options:
>     * make `appName` drive the [`STORAGE_NAME`](https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/contents.ts#L16)
>     * have a new key, e.g. `localStorageName`
>     * some combination of the two (prefer `localStorageName`, fall back to `appName`)
* generate an unique id during build that drives the [`STORAGE_NAME`](https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/contents.ts#L16) 

Pros:
- This would solve the issue of client's getting stuck on old `files` versions, regardless of the updates rolled out in the meantime.

Cons:
- After an update, clients would immediately lose access to local changes.
  -  This could be properly handled as the old content is still somewhere in the `indexedDB`.
  
If a sequential id generator is used, this could even provide some kind of version control on a file system level.

**Comment by bollwyvl** [comment]
> generate an unique id

mmm... smells like magic.

> If a sequential id generator

At some point, we'd probably just add `git` or `fossil` provider for #315 rather than re-inventing it poorly.

I'll go forward with the simplest addition of a `localStorageName` entry for `jupyter-lite.json`. This would give a site deployer the ability to override it if they don't want the default behavior, but not change the existing behavior too much.

**Comment by bollwyvl** [comment]
Started this up over here https://github.com/jupyterlite/jupyterlite/pull/449

Getting down in the code, I noted we need to also consider settings, which are stored in-browser: I've made them separate for now, as I could imagine the case (e.g. a series of tutorials) where a deployer has no opinion on settings being shared (e.g. I've configured a larger font so i can read more easily) but wouldn't want the contents shared.
